From 242cec703bbae5ac33862db483e7a9c968deb133 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20Br=C3=BCns?= <stefan.bruens@rwth-aachen.de>
Date: Fri, 3 Apr 2026 21:59:14 +0200
Subject: [PATCH] Fix required C++ version for CMake builds

Current GTest versions required C++14 or C++17, so set the appropriate
version in the compile_features.

Also change the library from using CMAKE_CXX_STANDARD to the equivalent
target_compile_features. This avoids downgrading the used standard
version and is properly exported to CMake Target.
---
 CMakeLists.txt | 23 +++++++++++------------
 1 file changed, 11 insertions(+), 12 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c23d634..9d853c9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,15 +8,6 @@ cmake_minimum_required(VERSION 3.5...4.0)
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake") # custom CMake modules like FindSQLiteCpp
 project(SQLiteCpp VERSION 3.3.3)
 
-# SQLiteC++ 3.x requires C++11 features
-if (NOT CMAKE_CXX_STANDARD)
-    set(CMAKE_CXX_STANDARD 11)
-elseif (CMAKE_CXX_STANDARD LESS 11)
-    message(WARNING "CMAKE_CXX_STANDARD has been set to '${CMAKE_CXX_STANDARD}' which is lower than the minimum required standard (c++11).")
-endif ()
-message(STATUS "Using c++ standard c++${CMAKE_CXX_STANDARD}")
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-
 message (STATUS "CMake version: ${CMAKE_VERSION}")
 message (STATUS "Project version: ${PROJECT_VERSION}")
 
@@ -212,6 +203,8 @@ endif()
 
 # add sources of the wrapper as a "SQLiteCpp" static library
 add_library(SQLiteCpp ${SQLITECPP_SRC} ${SQLITECPP_INC} ${SQLITECPP_DOC} ${SQLITECPP_SCRIPT})
+# SQLiteC++ 3.x requires C++11 features
+target_compile_features(SQLiteCpp PUBLIC cxx_std_11)
 
 # Options relative to SQLite and SQLiteC++ functions
 
@@ -475,10 +468,15 @@ if (SQLITECPP_BUILD_TESTS)
     target_link_libraries(SQLiteCpp_tests SQLiteCpp)
 
     find_package(GTest)
-    if (GTEST_FOUND)
+    if (GTest_FOUND)
         message(STATUS "Link to GTest system library")
         target_link_libraries(SQLiteCpp_tests GTest::GTest GTest::Main)
-    else (GTEST_FOUND)
+        if (GTest_VERSION AND (GTest_VERSION VERSION_LESS 1.17))
+            target_compile_features(SQLiteCpp_tests PRIVATE cxx_std_14)
+        else()
+            target_compile_features(SQLiteCpp_tests PRIVATE cxx_std_17)
+        endif()
+    else (GTest_FOUND)
         message(STATUS "Compile googletest from source in submodule")
         # deactivate some warnings for compiling the googletest library
         if (NOT MSVC)
@@ -505,7 +503,8 @@ if (SQLITECPP_BUILD_TESTS)
         endif (MSVC)
 
         target_link_libraries(SQLiteCpp_tests gtest_main)
-    endif (GTEST_FOUND)
+        target_compile_features(SQLiteCpp_tests PUBLIC cxx_std_17)
+    endif (GTest_FOUND)
 
     # add a "test" target:
     enable_testing()
-- 
2.53.0

